### PR TITLE
Fixed issue #2299: Better support the git staging process

### PR DESCRIPTION
**Pull request by tchesftw**
This commit adds staging support in the Commit window, including hunk and line staging. It also fixes a minor bug in the handling of renamed files in the View Patch dialog.

The window has a new checkbox: "Staging support". Leaving it unchecked, it all should behave exactly as it used to. If it doesn't, it's a bug. Checking it immediately changes the state of the checkboxes for each file to reflect its staging status: checked means all the changes for the file are staged, unchecked means all changes unstaged, and indeterminate (gray square) means that only some changes for the file are staged. Clicking the checkboxes stages/unstages the whole file.

Checking "Staging support" also replaces "Show Patch>>" by two new links: "Partial Staging>>" and "Partial Unstaging>>". Clicking those makes it possible to select and stage/unstage hunks and lines in the View Patch window, which I reused for that purpose. It relies on Scintilla's lexer for figuring out which hunks or lines are selected.

I've done only limited manual testing. It probably needs fixes for cases such as amends, submodules and conflicts, along with performance improvements, test cases and the like. But, hopefully, this is a start.

### CODE REVIEW

**Comment by csware** [comment]
Thank you for this PR!

I'll have a closer look as soon as I have some more spare time.

**Comment by csware** [comment]
Amazing work! I haven't looked deeply into bit, but it's a hunge PR!

Can you also please fix warning?

**Comment by tchesftw** [comment]
> Amazing work! I haven't looked deeply into it, but it's a hunge PR!
> 
> Can you also please fix warning?

Take your time! Warnings are fixed.

In the meantime, I'm already using the feature, so, should I notice any more problems with it, I'll be fixing them when I can.

**Comment by asmwarrior** [comment]
Any good news about this pull request? Thanks.

**Comment by tchesftw** [comment]
> Any good news about this pull request? Thanks.

For what I originally set out to do, I would consider it mostly done, except for the performance issues still to be fixed, which affect only the "staging enabled" mode.

It does not do (yet) any handling of the shell overlay icons and the fact that the commit window still complains when deleted files have been added to the index.

I don't know whether TGit's maintainer has been able to review the PR yet.

**Comment by csware** [comment]
I'm sorry, I don't have much time I can dedicate to TortoiseGIt right now.

**Comment by csware** [comment]
I had a first look today.

Looks really nice!

And yes, performance needs to be improved...

**Comment by csware** [comment]
Still missing is support for staged files that are not in the working tree any more

File: src/TortoiseProc/DiffLinesForStaging.cpp [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software Foundation,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+//
+#include "stdafx.h"
+#include "DiffLinesForStaging.h"
+#include <regex>
+
+CDiffLinesForStaging::CDiffLinesForStaging(const std::unique_ptr<char[]>& text, int numLines, int firstLineSelected, int lastLineSelected)
```
Comment (csware) [inline src/TortoiseProc/DiffLinesForStaging.cpp:24]: Please don't use unique_ptr<>&, better directly use char* here

File: src/TortoiseProc/DiffLinesForStaging.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+		else if (ptr[i] == '\r' || ptr[i] == '\n')
+			eol_len = 1;
+		else
+		{
+			i++;
+			continue;
+		}
+
+		int linebuflen = i - last_i + eol_len + 1;
+		auto line = std::make_unique<char[]>(linebuflen);
```
Comment (csware) [inline src/TortoiseProc/DiffLinesForStaging.cpp:50]: Is a copy needed, what about a string_view?

File: src/TortoiseProc/DiffLinesForStaging.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+	for (size_t i = 0; i < linesCount; i++)
+		documentLength += m_linevec.at(i).size;
+	return documentLength;
+}
+
+// Takes a buffer containing the first line of a hunk (@@xxxxxx@@)
+// Parses it to extract its old lines count and new lines count and passes them back to the given oldCount and newCount.
+// Returns true if the line matches the expected format, false otherwise (what should never happen).
+// If allowSingleLine is false, returns false for hunks missing the start line number in one or both sides (e.g. @@ -x +y,z @@)
+bool CDiffLinesForStaging::GetOldAndNewLinesCountFromHunk(const std::unique_ptr<char[]>& strHunkStart,
```
Comment (csware) [inline src/TortoiseProc/DiffLinesForStaging.cpp:225]: Please don't use unique_ptr<>&, better directly use char* here

File: src/TortoiseProc/resource.h [hunk 4]
```diff
@@ -1,4 +1,4 @@
-//{{NO_DEPENDENCIES}}
+﻿//{{NO_DEPENDENCIES}}
```
Comment (csware) [inline src/TortoiseProc/resource.h:1]: Please don't change file encoding here

File: src/TortoiseProc/CommitDlg.cpp [hunk 5]
```diff
@@ -1352,7 +1394,10 @@ UINT CCommitDlg::StatusThread()
 	DialogEnableWindow(IDC_CHECKFILES, false);
 	DialogEnableWindow(IDC_CHECKSUBMODULES, false);
 
-	g_Git.RefreshGitIndex();
+	// When a file is completely staged by using the partial staging functionality (i.e., applying patches to the index),
+	// instead of clicking the checkbox (i.e., calling git add), Git may wrongly think that the file is partially staged
+	// if git update-index is called with --refresh instead of --really-refresh
```
Comment (csware) [inline src/TortoiseProc/CommitDlg.cpp:1399]: I'm sorry, I don't understand this ATM.

File: src/TortoiseProc/CommitDlg.cpp [hunk 6]
```diff
@@ -1478,8 +1524,8 @@ UINT CCommitDlg::StatusThread()
 				GitRev headRevision;
 				if (headRevision.GetParentFromHash(hash))
 					MessageBox(headRevision.GetLastErr(), L"TortoiseGit", MB_ICONERROR);
-				// do not allow to show diff to "last" revision if it has more that one parent
-				if (headRevision.ParentsCount() != 1)
+				// do not allow to show diff to "last" revision if it has more that one parent or staging support is enabled
+				if (headRevision.ParentsCount() != 1 || m_bStagingSupport)
```
Comment (csware) [inline src/TortoiseProc/CommitDlg.cpp:1528]: * Why is this needed?
* "Diff to last commit" is not automatically unchecked when staging support is activated.

File: src/TortoiseProc/CommitDlg.cpp [hunk 7]
```diff
@@ -744,148 +772,220 @@ void CCommitDlg::OnOK()
 		m_ListCtrl.WriteCheckedNamesToPathList(m_selectedPathList);
 	m_pathwatcher.Stop();
 	InterlockedExchange(&m_bBlock, TRUE);
-	//first add all the unversioned files the user selected
-	//and check if all versioned files are selected
+
 	int nchecked = 0;
 
-	CMassiveGitTask mgtReAddAfterCommit(L"add --ignore-errors -f");
+	CMassiveGitTask mgtReAddAfterCommit(L"add --ignore-errors -f"); // these two commands will not be executed if staging support is enabled
```
Comment (csware) [inline src/TortoiseProc/CommitDlg.cpp:778]: Better write this comment to all commands it applies to

File: src/TortoiseProc/CommitDlg.cpp [hunk 8]
```diff
@@ -2715,34 +2808,79 @@ void CCommitDlg::OnScnUpdateUI(NMHDR * /*pNMHDR*/, LRESULT *pResult)
 
 void CCommitDlg::TogglePatchView()
```
Comment (csware) [inline src/TortoiseProc/CommitDlg.cpp:2809]: Can we drop this method?

File: src/TortoiseProc/CommitDlg.h [hunk 9]
```diff
@@ -205,6 +232,10 @@ class CCommitDlg : public CResizableStandAloneDialog, public CSciEditContextMenu
 	CRect				m_LogMsgOrigRect;
 	CPathWatcher		m_pathwatcher;
 	CHyperLink			m_ctrlShowPatch;
+	CHyperLink			m_ctrlPartialStaging;
+	CHyperLink			m_ctrlPartialUnstaging;
+	bool				m_bPartialStagingTextCurrentlyIsShow; // false if currently "Hide Staging"
```
Comment (csware) [inline src/TortoiseProc/CommitDlg.h:237]: Do we need these two variables? Can't we just have one (e.g., an enum)?

**Review by csware — CHANGES_REQUESTED** [review]
There is a lot of copying strings, is that needed? Can't we just cache it when we load the patch?

File: src/TortoiseProc/CommitDlg.cpp (hunk 6 above)
Comment (tchesftw) [reply src/TortoiseProc/CommitDlg.cpp:1534]: > "Diff to last commit" is not automatically unchecked when staging support is activated.

That's right. Only when doing an amend, "Show diff to last commit" must stay checked and grayed out.

> Why is this needed?

Comment clarified.

File: src/TortoiseProc/DiffLinesForStaging.cpp (hunk 1 above)
Comment (csware) [reply src/TortoiseProc/DiffLinesForStaging.cpp]: You got me wrong here. Just use `char*` here and don't pass a const reference to a `unique_ptr`.

**Comment by csware** [comment]
Thank you for all the work you have done here!

File: test/UnitTests/StagingTest.cpp [hunk 10]
```diff
@@ -21,16 +21,16 @@
...
 {
 	CString resourceDir;
 	GetResourcesDir(resourceDir);
 
 	CFile file;
 	file.Open(resourceDir + resourceFile, CFile::modeRead | CFile::typeBinary);
 	auto filelen = file.GetLength();
-	auto buf = std::make_unique<char[]>(static_cast<size_t>(filelen) + 1);
-	file.Read(buf.get(), static_cast<UINT>(filelen));
+	auto buf = new char[static_cast<size_t>(filelen) + 1]();
```
Comment (csware) [inline test/UnitTests/StagingTest.cpp:32]: Here:
* Is this copy of the string realyl necessary?
* Please keep using the `unique_ptr` here.

**Comment by csware** [comment]
I still don't undertand why `GetFullLineByLineNumber` needs to return a copy of the string and not just a stringview on the whole buffer?

**Comment by csware** [comment]
When all these copies are really needed, why no `std::string`? That could save us lots of ptr+length pairs

**Comment by tchesftw** [comment]
I changed the places where `std::unique_ptr<char[]>&` was used to just `char*`

The tests are taking a few seconds to run.


> I still don't undertand why `GetFullLineByLineNumber` needs to return a copy of the string and not just a stringview on the whole buffer?

Because `ParseHunkOnEitherSelectionBoundary` may need to change the first character of the line, i.e. when turning a +/- line into a context line.



> When all these copies are really needed, why no `std::string`? That could save us lots of ptr+length pairs

Mostly a legacy from an early implementation I didn't publish. It's costly to change it now because it would require changing the way an invalid line number is recognized, from returning a null pointer (such as currently done) to doing something else.

**Comment by csware** [comment]
Refactoring is nearly finished, then it's ready to merge!

File: src/TortoiseProc/CommitDlg.h [hunk 11]
```diff
@@ -228,13 +268,24 @@ class CCommitDlg : public CResizableStandAloneDialog, public CSciEditContextMenu
 	HACCEL				m_hAccel;
 	bool				RestoreFiles(bool doNotAsk = false, bool allowCancel = true);
 
+	// State of the links "Partial Staging>>" and "Partial Unstaging>>" shown on the commit window
+	enum PartialStagingDisplayState : unsigned char
```
Comment (csware) [inline src/TortoiseProc/CommitDlg.h:272]: I still wonder why we use a bitset here as one needs to be enabled, right?

**Comment by csware** [comment]
Have you also tested the rebase dialog?

File: src/TortoiseProc/CommitDlg.h (hunk 11 above)
Comment (tchesftw) [reply src/TortoiseProc/CommitDlg.h:272]: Both can be set at the same time (see the constructor of `CCommitDlg`), but never unset at the same time.

When either one is set, its corresponding link displays "Partial (un)staging". When either one is unset, its corresponding link displays "Hide (un)staging". So, if both are set, the View Patch window is closed and the user can click either one of them to display the partial staging or unstaging window.

Comment (csware) [reply src/TortoiseProc/CommitDlg.h:272]: Yeah, I see.

**Comment by csware** [comment]
Could you please have a look at my latest commit? I added support for amending. To me it looks good, but it would be great if you could doublecheck.

**Review by csware — APPROVED** [review]

**Comment by csware** [comment]
Finally, I also have a version which tires to allocate as less memory as possible by using `string_view`

**Comment by tchesftw** [comment]
> Finally, I also have a version which tires to allocate as less memory as possible by using `string_view`

Interesting. I had tried to do that but it turned out a mess. You seem to have just replaced a nullptr with an empty string. I thought it would cause problems, but the tests are succeeding and they are designed to catch this kind of thing. Great, then!



> Could you please have a look at my latest commit? I added support for amending. To me it looks good, but it would be great if you could doublecheck.
> 
> Using it one can now unstage (and stage) changes when amending.

I will.

**Comment by tchesftw** [comment]
> Could you please have a look at my latest commit? I added support for amending. To me it looks good, but it would be great if you could doublecheck.
> 
> Using it one can now unstage (and stage) changes when amending.

Seems to work nicely. I hadn't thought of it.
